### ISSUE TITLE
Using Bazel as a build tool

### ISSUE THREAD

**Issue by JesseFarebro** [google/xi-editor#45]
I know the project is in its infancy, but considering the possibility it scales up to having multiple front-ends (GTK3, Cocoa, Win32, ncurses) building the project could get out of hand.

[Bazel](http://bazel.io) could solve these issues. Right now there are Skylark rules for [building rust](https://github.com/bazelbuild/rules_rust) libraries/binaries.

I'm not overly familiar with the [Objective-C rules](http://bazel.io/docs/be/objective-c.html) in Bazel but it could be something to look at for building the OS X front-end.

As far as building other front-ends, other GUI toolkits seem to be a little more forgiving and will have bindings available for popular languages which Bazel is capable of compiling.

I'd love to get the discussion going right now and I could whip up a quick WIP of a Bazel build if this is something you'd like to consider.

**Comment by gaganjyot** [comment]
Hi,
also my 2 cents,

Instead of actually adding the Ui code here itself in the main repository, I'd love to see it as different repositories. one for xi-cocoa, one for xi-GTK, and one for xi-windows. I personally feel this will be good is because this way the engine will be totally separate from the GUI in all terms and developers will be concerned with only the thing they will be working on. Once the json specification for transferring information  between frontend and backend is created well, the developers could just see that specification and implement the parts in backend and frontends respectively. Also this way ( Just a thought, I see xi scaling up high ), if someone likes the front end that is created by Xi developers but prefers some other backend, they could easily use the other one if the other backend software also uses the same json protocol.
P.S I like the idea of Xi very much but at the moment I'm in Summer of Code, I'd love to contribute to Xi after the GSoC. :)

**Comment by raphlinus** [comment]
I'm open to Bazel as the build tool, but I have concerns. How many additional steps does it add to the "new developer" ramp-up? Does working on the Mac front-end feel as "native" as the current Xcode experience? Is Windows support solid enough? Currently the [Bazel FAQ](http://bazel.io/faq.html) says: *"We are currently actively working on improving Windows support, but it's still ways from being usable."*

I'm open to the idea of separate repositories, but for now it's convenient having just one, as the FE/BE protocol is revving rapidly and a single repository makes it possible to do a change to it atomically. At some point the protocol will stabilize (and changes perhaps protected by version negotiation), at which more loosely coupled projects would make more sense.

**Comment by rtsuk** [comment]
Swift is a P2 priority for July. I think it would be good to reconsider a cross platform build system after that.

The first step should be to come up with a list of desired features of this build system and get some consensus on them. For example, I wouldn't be happy with a system that didn't generate clean, fast Xcode project files.

http://bazel.io/roadmap.html

**Comment by gaganjyot** [comment]
Yes, Even I meant to separate the FE/BE after the stabilization of the protocol. :) High hopes from Xi.

**Comment by frewsxcv** [comment]
In what ways is Cargo insufficient as a build tool such that third-party build tools are necessary?
